python/__init__.py

```diff
@@ -42,33 +50,73 @@ def pack_can_buffer(arr):
     header[2] = (word_4b >> 8) & 0xFF
     header[3] = (word_4b >> 16) & 0xFF
     header[4] = (word_4b >> 24) & 0xFF
-    snds[idx] += header + dat
-    if len(snds[idx]) > 256: # Limit chunks to 256 bytes
-      snds.append(b'')
-      idx += 1
-
-  #Apply counter to each 64 byte packet
-  for idx in range(len(snds)):
-    tx = b''
-    counter = 0
-    for i in range (0, len(snds[idx]), 63):
-      tx += bytes([counter]) + snds[idx][i:i+63]
+    msg = header + dat
+
+    overflow_len = max(0, len(msg) - (DATA_SIZE_PER_CHUNK - len(chunks[-1])))
+    if overflow_len == 0:
+      # we have enough space left for this message
+      chunks[-1] += msg
+    else:
+      # add partial data, and overflow to next chunk
+      chunks[-1] += msg[:len(msg) - overflow_len]
+      chunks[-1] = CAN_CHUNK_HEADER_STRUCT.pack(counter, prev_overflow_len, len(chunks[-1])) + chunks[-1]
+      chunks.append(msg[len(msg) - overflow_len:])
+      prev_overflow_len = overflow_len if overflow_len < len(msg) else 0
       counter += 1
-    snds[idx] = tx
-  return snds
+      counter %= 256
+
+    # split up chunks into roughly 256 byte transfers with no overflow
+    if len(chunks) >= (256 / CHUNK_SIZE):
+      # finish last header
+      chunks[-1] = CAN_CHUNK_HEADER_STRUCT.pack(counter, prev_overflow_len, len(chunks[-1])) + chunks[-1]
+      counter += 1
+      counter %= 256
+
+      # commit
+      transfers.append(b"".join(chunks))
+      chunks = [b""]
+      prev_overflow_len = 0
+
+  # finish last chunk header for this transfer and commit last chunks
+  chunks[-1] = CAN_CHUNK_HEADER_STRUCT.pack(counter, prev_overflow_len, len(chunks[-1])) + chunks[-1]
+  transfers.append(b"".join(chunks))
+
+  return transfers
 
-def unpack_can_buffer(dat):
+def unpack_can_buffer(dat, prev_rx_counter=None):
   ret = []
-  counter = 0
   tail = bytearray()
```

Comment:
Resetting the tail prevents having overflow data between different unpack buffers (should never happen on USB). We just discard this data for now, might still want to fix this for the SPI implementation